### ISSUE TITLE
steps to achieve accuracy

### ISSUE THREAD

**Issue by benam2** [hugochan/KATE#23]
Could you explain which scripts need to be run to feed the output of encoder to the classifier and finally calculate the accuracy?
you have two files, `run_clf` and `run_classifier`.

**Comment by hugochan** [comment]
@benam2 You can run the run_clf.py file.

**Comment by benam2** [comment]
Could you please explain why did you use `Sigmoid` rather than `Softmax`?
One more doubt; you have split train/test manually, am I right? 
Is there any reason for that or because of the problem of not seeing some classes in the test phase you have done that?

**Comment by hugochan** [comment]
@benam2 I randomly split the whole dataset as train/valid/test for training/validation/evaluation. Could you please report your results here?

**Comment by benam2** [comment]
Sure, but I think Im doing something wrong!!!
let me check this first:

where did you calculate val_doc_labels = load_json(args.val_doc_labels)` and `test_doc_codes = load_json(args.test_doc_codes)` which is in `run_clf` file?

For `val_doc_labels` I did like this: `generate_20news_doc_labels(val_doc_codes.keys(), os.path.join(out_dir, 'val.labels'))`

And in case of `test_doc_codes`, I added `dump_json(dict(zip(test_doc_keys.tolist(), test_doc_codes.tolist())),
          './csvfiles/'+ dir +'/output.test')`. ( I did not find where that you have calculated these).

Is it okay in this way?

**Comment by benam2** [comment]
Could you please help with the above question?

**Comment by hugochan** [comment]
@benam2 To reproduce the results, you can do the following after training the model:

(1) python pred.py -i [input_data] -lm [model_file] -o [output_doc_vec_file]
You will get a doc_codes file for training data when input_data is set to the training dir and a doc_codes file for testing data when input_data is set to the testing dir.

(2) python run_classifier.py [train_doc_codes] [train_doc_labels]  [test_doc_codes] [test_doc_labels] -nv [num_val] -ne [num_epoch] -bs [batch_size]

Note the difference between run_clf.py and run_classifier.py is that the later randomly selects samples from training data for validation.

**Comment by benam2** [comment]
Thank you for helping me on this, I have not done this before (Applying classifier over the encoder so forgive me if my question sounds simple).

Actually, I did the same except instead of preparing the data via the code available in the  `pred.py` script, I used the same code in `train.py` as they are identical: I'm talking this part of the code:

`if args.output:
        train_doc_codes = ae.encoder.predict(X_train)
        val_doc_codes = ae.encoder.predict(X_val)
        doc_keys = np.array(doc_keys)
        dump_json(dict(zip(doc_keys[train_idx].tolist(), train_doc_codes.tolist())), args.output + '.train')
        dump_json(dict(zip(doc_keys[val_idx].tolist(), val_doc_codes.tolist())), args.output + '.val')
        print 'Saved doc codes file to %s and %s' % (args.output + '.train', args.output + '.val')`

I did this for both training and test data.

The only difference I can see is that here we use ` ae.encoder.predict` on the train and test data, however, in `pred.py` it is using the whole model, not encoder [doc_codes = ae.predict(X_docs)](url).

I think the second one in the `train.py` script is more correct, Isn't that right? we should get prediction over the `encoder` not the whole `autoencoder` model.

Please correct me if I'm wrong
Thanks~

**Comment by benam2** [comment]
So per your question, I got 71% accuracy doing with the approach I explained above.

**Comment by hugochan** [comment]
> Thank you for helping me on this, I have not done this before (Applying classifier over the encoder so forgive me if my question sounds simple).
> 
> Actually, I did the same except instead of preparing the data via the code available in the `pred.py` script, I used the same code in `train.py` as they are identical: I'm talking this part of the code:
> 
> `if args.output: train_doc_codes = ae.encoder.predict(X_train) val_doc_codes = ae.encoder.predict(X_val) doc_keys = np.array(doc_keys) dump_json(dict(zip(doc_keys[train_idx].tolist(), train_doc_codes.tolist())), args.output + '.train') dump_json(dict(zip(doc_keys[val_idx].tolist(), val_doc_codes.tolist())), args.output + '.val') print 'Saved doc codes file to %s and %s' % (args.output + '.train', args.output + '.val')`
> 
> I did this for both training and test data.
> 
> The only difference I can see is that here we use ` ae.encoder.predict` on the train and test data, however, in `pred.py` it is using the whole model, not encoder `doc_codes = ae.predict(X_docs)]`.
> 
> I think the second one in the `train.py` script is more correct, Isn't that right? we should get prediction over the `encoder` not the whole `autoencoder` model.
> 
> Please correct me if I'm wrong
> Thanks~

Note that in pred.py it is using the encoder part since only the encoder part is saved during training, see [here](https://github.com/hugochan/KATE/blob/master/autoencoder/core/ae.py#L89).

**Comment by benam2** [comment]
Sorry, I should have missed that part!
I have a general question

**Comment by benam2** [comment]
Iam still getting 61% acc 
this is the way I have run it:

here stop_words = true
stem = True
`python construct_20news.py -train /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/train -test /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/test -o ./output -threshold 5 -topn 2000`

`python train.py -i ./output/train.corpus -nd 128 -ne 100 -bs 100 -nv 1000 -ctype kcomp -ck 32 -sm ./output/model`

`python pred.py -i ./output/train.corpus -lm ./output/model -o ./output/docoutput_train -st ./output/st`

`python pred.py -i ./output/test.corpus -lm ./output/model -o ./output/docoutput_test -st ./output/st`

`python run_classifier.py ./output/docoutput_train ./output/train.labels ./output/docoutput_test ./output/test.labels  -cv 5`

I want to get the same result as reported in the paper, please let me know which parameter you want me to change.

**Comment by hugochan** [comment]
> Iam still getting 61% acc
> this is the way I have run it:
> 
> here stop_words = true
> stem = True
> `python construct_20news.py -train /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/train -test /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/test -o ./output -threshold 5 -topn 2000`
> 
> `python train.py -i ./output/train.corpus -nd 128 -ne 100 -bs 100 -nv 1000 -ctype kcomp -ck 32 -sm ./output/model`
> 
> `python pred.py -i ./output/train.corpus -lm ./output/model -o ./output/docoutput_train -st ./output/st`
> 
> `python pred.py -i ./output/test.corpus -lm ./output/model -o ./output/docoutput_test -st ./output/st`
> 
> `python run_classifier.py ./output/docoutput_train ./output/train.labels ./output/docoutput_test ./output/test.labels -cv 5`
> 
> I want to get the same result as reported in the paper, please let me know which parameter you want me to change.

Could you please try the following settings:
1) set batch size to 50 when training KATE on 20News
2) please do not use cross-validation (turn off -cv when running run_classifier.py) because other wise you will need to do cross-validation for training KATE as well. The principle is that you need to keep the same data split for training KATE and the classifier.
3) set batch size to 50 when training the classifier.

**Comment by benam2** [comment]
Ok I will try this way and report you Accuracy.

**Comment by benam2** [comment]
I just tried and nothing changed. I also faced with early stopping. To make sure I have not changed anything I downloaded your source code again.

The only thing that you have not mentioned in the doc is about splitting data. because in the construction phase you make test and train vocab separately. For that reason, I split data as 30 test and 70 train on my own and then followed the instruction you have shared.
So this is what I run this time:

```
python construct_20news.py -train /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/train -test /home/sgnbx/Downloads/projects/LSTM_autoencoder/Data/20news/test -o ./output -threshold 5 -topn 2000

python train.py -i ./output/train.corpus -nd 128 -ne 100 -bs 50 -nv 1000 -ctype kcomp -ck 32 -sm ./output/model

python pred.py -i ./output/train.corpus -lm ./output/model -o ./output/docoutput_train -st ./output/st

python pred.py -i ./output/test.corpus -lm ./output/model -o ./output/docoutput_test -st ./output/st

python run_classifier.py ./output/docoutput_train ./output/train.labels ./output/docoutput_test ./output/test.labels -nv 1000
```
in `run_classifier` I tried with and without `nv 1000`.

I am open to change anything to make sure I have not missed anything.

**Comment by benam2** [comment]
1. I havent tried DR yet, 
2. I'm gonna give the Reuters and wiki a try: the link in the paper fro the Reuters has the vectors of the documents, could you share which link did you use?
3. Actually Keras and TF version is different, but I am doubtful only version can have 9 percent acc decline, But I will give it a try by creating another env with the exact version.

**Comment by benam2** [comment]
I appreciate it if you give me the link that you used for the reuters, so we will be in the sam e setting(the one in the paper did not have a file for label so that later we can calculate accuracy..., it sounds the provided code needed a list of path files.

One more quick question as well, could you share your experience running `train_vae`? which parameters did you try with `n_dim and k comp`? I don't get why in the code kcomp is being called like this `kcomp[1]`?

**Comment by hugochan** [comment]
> I appreciate it if you give me the link that you used for the reuters, so we will be in the sam e setting(the one in the paper did not have a file for label so that later we can calculate accuracy..., it sounds the provided code needed a list of path files.
> 
> One more quick question as well, could you share your experience running `train_vae`? which parameters did you try with `n_dim and k comp`? I don't get why in the code kcomp is being called like this `kcomp[1]`?

I just uploaded the preprocessed data I used in experiments [here](http://academic.hugochan.net/download/KATE-preprocessed-data.zip). I do not have the preprocessed Reuters data on my machine now since it is too large. I will check if I have that on my external disk this evening. 

You do not need to set ctype and comp_topk when running VAE (which I tried and it did not work with VAE), you just need to run the regular VAE. The second hidden layer size of VAE should match the hidden size of KATE, I forget what was the exact value I used for the first hidden layer size (you can tune it in your experiments).

**Comment by benam2** [comment]
I really appreciate it if you can share the reuters as well. I have the data set but it different setting from the one in the paper.

Got it, I thought you were able to run VAE using the competitive layer, thanks for the clarification.

**Comment by hugochan** [comment]
> I really appreciate it if you can share the reuters as well. I have the data set but it different setting from the one in the paper.
> 
> Got it, I thought you were able to run VAE using the competitive layer, thanks for the clarification.

Unfortunately, I do not have the preprocessed reuters now. However, you should be able to preprocess the data on your own using our scripts. Here is the link to the raw [Reuters](http://www.ai.mit.edu/projects/jmlr/papers/volume5/lewis04a/lyrl2004_rcv1v2_README.htm) data. You can preprocess it as follows:
1) Download the raw data (i.e., 1 training data file and 4 testing data files) from the above website.
2) Preprocess the raw data by running construct_reuters.py
3) Download [rcv1-v2.topics.qrels.gz](http://www.ai.mit.edu/projects/jmlr/papers/volume5/lewis04a/a08-topic-qrels/rcv1-v2.topics.qrels.gz) to get the document-label file.
4) Extract document labels by running get_reuters_labels.py

**Comment by benam2** [comment]
I run your code on wiki and movie. I got the same result as reported for the WIKI.
My system has only 32 GB RAM so I faced with Memory error to run Reuters data set (in np.r_). 
I am trying to figure out which options of AWS Amazon service is sufficient to get Reuters run there.
I think having 40 GB RAM should be enough (data itself needs 20GB ram, however for some reason it probably does not delete `del docs[k]` and that's why it needs more RAM, Im not sure though)
